test/org/javarosa/core/model/Safe2014DagImplTest.java

```diff
@@ -0,0 +1,396 @@
+
+        assertThat(repeats.size(), equalTo(4));
+        assertThat(repeats.get(0).getChildAt(0).getValue().getDisplayText(), equalTo("A"));
+        assertThat(repeats.get(1).getChildAt(0).getValue().getDisplayText(), equalTo("B"));
+        assertThat(repeats.get(2).getChildAt(0).getValue().getDisplayText(), equalTo("D"));
+        assertThat(repeats.get(3).getChildAt(0).getValue().getDisplayText(), equalTo("E"));
+
+        assertThat(summaryNode.getValue().getDisplayText(), equalTo("ABDE"));
+
+        // check that correct calculations were triggered
+        final String[] expectedMessages = {
+                "Processing 'Deleted: houseM [3]: 0 triggerables were fired.' for ",
+                "Processing 'Recalculate' for summary [1] (ABDE)",
+                "Processing 'Deleted: name [3_1]: 1 triggerables were fired.' for ",
+                "Processing 'Deleted: houseM [4]: 0 triggerables were fired.' for "
+        };
+
+        assertThat(dagEvents.size(), equalTo(expectedMessages.length));
+
+        int messageIndex = 0;
+        for (String expectedMessage : expectedMessages) {
+            assertThat(dagEvents.get(messageIndex++).getDisplayMessage(), equalTo(expectedMessage));
+        }
+    }
+
+    @Test
+    public void deleteRepeatGroupWithCalculationsTimingTest() throws Exception {
+        // Given
+        final FormDef formDef =
+                parse(r("delete-repeat-group-with-calculations-timing-test.xml")).formDef;
+
+        assertIDagImplUnderTest(formDef);
+
+        formDef.initialize(false, new InstanceInitializationFactory()); // trigger all calculations
+
+        final FormInstance mainInstance = formDef.getMainInstance();
+
+        // Construct the required amount of repeats
+        final TreeElement templateRepeat = mainInstance.getRoot().getChildAt(0);
+        final int numberOfRepeats = 1000;
+        for (int i = 0; i < numberOfRepeats; i++) {
+            final TreeReference refToNewRepeat = templateRepeat.getRef();
+            refToNewRepeat.setMultiplicity(1, i); // set the correct multiplicity
+
+            final FormIndex indexOfNewRepeat = new FormIndex(0, i, refToNewRepeat);
+            formDef.createNewRepeat(indexOfNewRepeat);
+        }
+
+        final TreeElement firstRepeat = mainInstance.getRoot().getChildAt(1);
+        final TreeReference firstRepeatRef = firstRepeat.getRef();
+        final FormIndex firstRepeatIndex = new FormIndex(0, 0, firstRepeatRef);
+
+        // When
+        long startMs = System.currentTimeMillis();
+
+        for (int i = 0; i < numberOfRepeats; i++) {
+            long currentIterationStart = System.currentTimeMillis();
+            formDef.deleteRepeat(firstRepeatIndex);
+            long took = System.currentTimeMillis() - currentIterationStart;
+            String tookFormatted = LocalTime.fromMillisOfDay(took).toString();
+            System.out.println("Deletion of repeat #" + i + " took " + tookFormatted);
+        }
+
```

Comment:
Just a thought. If you do something like this:
```
System.out.println("Repeat Group\nDeletion Time (ms)");
for (int i = 0; i < numberOfRepeats; i++) {
    long currentIterationStart = System.nanoTime();
    formDef.deleteRepeat(firstRepeatIndex);
    double tookMs = (System.nanoTime() - currentIterationStart) / 1000000D;
    System.out.printf("%d\t%.3f\n", i, tookMs);
}
```
Then we can do neat things like this:

<img width="697" alt="deletion times" src="https://user-images.githubusercontent.com/55099/33785488-185580f0-dc1a-11e7-8170-87e91bf0f7cd.png">

Reply:
@dcbriccetti Thanks, I'll fix the output to match this format.

@lognaturel 
If the deletion times are acceptable then I'd like to clean up the test so it can be merged. The number of iterations should be decreased to not slow down build times - what would say about lowering it to 200 elements? Deleting all repeats takes about 3.5 seconds in that case.

Reply:
@dcbriccetti @lognaturel I pushed the clean-up timing test. It uses format suggested by @dcbriccetti and tests deletion of 200 repeats.

Reply:
What we did for the timing “tests” in XFormParserTest was make them run very fast (a fraction of a second), but have a comment with instructions for anyone wanting to make them run longer, for actual timing purposes:
`        for (double powerOfTen = 3; powerOfTen <= 4.0; powerOfTen += 0.1) {  // Raise this upper limit to really measure
`

Reply:
@dcbriccetti Right, thanks! I added a comment and left 200 repeats. It runs fast and it's still quite a lot.